src/main/kotlin/com/exactpro/th2/ws/client/api/impl/WebSocketClient.kt

```diff
@@ -165,47 +167,49 @@ class WebSocketClient(
         connect()
     }
 
-    @Synchronized
-    fun start() = when {
-        isRunning -> onInfo { "Client is already running" }
-        else -> {
-            onInfo { "Starting client" }
-            isRunning = true
-            connect()
-            onInfo { "Started client" }
+    fun start() = lock.withLock {
+        when {
+            isRunning -> onInfo { "Client is already running" }
+            else -> {
+                onInfo { "Starting client" }
+                isRunning = true
+                connect()
+                onInfo { "Started client" }
+            }
         }
     }
 
-    @Synchronized
-    fun stop() = when {
-        !isRunning || !::socket.isInitialized -> onInfo { "Client is already stopped" }
-        else -> {
-            onInfo { "Stopping client" }
+    fun stop() = lock.withLock {
+        if (!isRunning || !::socket.isInitialized) {
+            onInfo { "Client is already stopped" }
+            return
+        }
 
-            socket.runCatching {
-                isRunning = false
+        onInfo { "Stopping client" }
 
-                if (isOutputClosed) {
-                    logger.warn { "Trying to close socket abruptly" }
-                    abort()
-                } else {
-                    logger.info { "Trying to close socket gracefully" }
+        socket.runCatching {
+            isRunning = false
 
-                    handler.runCatching(IHandler::preClose).onFailure {
-                        onError(it) { "Failed to handle preClose event" }
-                    }
+            if (isOutputClosed) {
+                logger.warn { "Trying to close socket abruptly" }
+                abort()
+            } else {
+                logger.info { "Trying to close socket gracefully" }
 
-                    sendClose(WebSocket.NORMAL_CLOSURE, "")
+                handler.runCatching(IHandler::preClose).onFailure {
+                    onError(it) { "Failed to handle preClose event" }
                 }
-            }.onFailure {
-                logger.error(it) { "Failed to close socket" }
-            }
 
-            onInfo { "Stopped client" }
+                sendClose(WebSocket.NORMAL_CLOSURE, "")
+            }
+        }.onFailure {
+            logger.error(it) { "Failed to close socket" }
         }
+
+        onInfo { "Stopped client" }
     }
 
-    private fun connect() {
+    private fun connect() = lock.withLock {
```

Comment:
`lock.withLock {` redundant because all public methods are already executed under the lock

Reply:
Not all of them. See `onClose` / `onError` method

Reply:
Maybe add the `lock.withLock` to all public methods

Reply:
There's no need to synchronize all of them (except to lift thread-safety requirement from handler)

Reply:
Also I'm not sure that that won't cause a deadlock in some cases